Use a postprocessing step to apply the pressure constraint in loops

GiudGiud opened the pull request:
This saves 1 degree of freedom. 
an important one though.

GiudGiud committed:
Add a user object to fix the pressure after each solve
refs #25446

GiudGiud committed:
Add pin parameters, get the pin offsets mostly
refs #25446

GiudGiud committed:
More implementation of the NSFV pin
- fixup parameters
- add parallel communication of point value
- add check on execute on flag for average value
refs #25446

GiudGiud committed:
Add offset of pressure value
refs #25446

GiudGiud committed:
Add UO pressure pin to NSFV action
Make it the default
refs #25446

GiudGiud committed:
Add testing to new pressure pin UO treatment
refs #25446

GiudGiud committed:
Fixup action logic
Use point value PP logic to find the point value
refs #25446

GiudGiud committed:
Add pressure pin UO doco, refs #25446

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1766019/) on f8d41e7 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25445/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1766061/) on f8d41e7 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>1acda2</th>
      <th colspan=4>#25445 f8d41e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.70%</td>
      <td>85.70%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>91495</td>
      <td>91491</td>
      <td>-4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15266</td>
      <td>15270</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25445/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25445/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>1acda2</th>
      <th colspan=4>#25445 f8d41e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.17%</td>
      <td>84.17%</td>
      <td>-0.00%</td>
      <td>83.53%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>11166</td>
      <td>11229</td>
      <td>+63</td>
      <td>71</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2100</td>
      <td>2112</td>
      <td>+12</td>
      <td>14</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25445/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25445/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25445/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 83.53% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
wow green on the first try

lindsayad commented on modules/navier_stokes/src/userobjects/NSFVPressurePin.C:
```diff
@@ -0,0 +1,127 @@
+  _sys.system().local_dof_indices(_p->number(), local_dofs);
+  for (const auto & dof : local_dofs)
+    sln.add(dof, pin_value);
+  sln.close();
```
```suggestion
  sln.close();
  _sys.system().update();
```
To be safe I would ensure the changes in the parallel vector get into the ghosted solution vector, since that is what is read by most objects

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/tests:
```diff
@@ -25,20 +32,36 @@
+    type = 'Exodiff'
+    input = 'lid-driven-action.i'
+    exodiff = 'lid-driven-point-pressure-action.e'
+    requirement = 'The system shall be able to solve the incompressible Navier-Stokes equations in a lid-driven cavity by fixing the point value of the pressure at a certain coordinate using the NSFV action syntax, with a post-treatment of the pressure rather than a Lagrange multiplier-based constraint.'
```
```suggestion
    requirement = 'The system shall be able to solve the incompressible Navier-Stokes equations in a lid-driven cavity by fixing the point value of the pressure at a certain coordinate using the Navier-Stokes finite volume action syntax, with a post-treatment of the pressure rather than a Lagrange multiplier-based constraint.'
```

GiudGiud commented on modules/navier_stokes/doc/content/source/userobjects/NSFVPressurePin.md:
```diff
@@ -0,0 +1,22 @@
+
+
+!alert note
+In the [NSFVAction.md], a `NSFVPressurePin` can be passed by the setting the [!param](/Modules/NavierStokesFV/pinned_pressure_type) parameter
```
```suggestion
In the [NSFVAction.md], a `NSFVPressurePin` can be used by setting the [!param](/Modules/NavierStokesFV/pinned_pressure_type) parameter
```

GiudGiud commented on modules/navier_stokes/include/userobjects/NSFVPressurePin.h:
```diff
@@ -0,0 +1,78 @@
+ * This user-object corrects the pressure
+ */
+class NSFVPressurePin : public GeneralUserObject,
+                        public TaggingInterface,
```
```suggestion
```

GiudGiud commented on modules/navier_stokes/include/userobjects/NSFVPressurePin.h:
```diff
@@ -0,0 +1,78 @@
+
+#pragma once
+
+#include "GeneralUserObject.h"
+#include "TaggingInterface.h"
+#include "BlockRestrictable.h"
+#include "ADReal.h"
+#include "MooseTypes.h"
+#include "CellCenteredMapFunctor.h"
+#include "VectorComponentFunctor.h"
+#include "FaceArgInterface.h"
+#include "INSFVPressureVariable.h"
+#include "ADFunctorInterface.h"
+
+#include "libmesh/vector_value.h"
+#include "libmesh/id_types.h"
+#include "libmesh/stored_range.h"
+#include <unordered_map>
+#include <set>
+#include <unordered_set>
+
+class MooseMesh;
+class INSFVVelocityVariable;
+class INSFVPressureVariable;
+namespace libMesh
+{
+class Elem;
+class MeshBase;
+}
```
```suggestion
#include "GeneralUserObject.h"
#include "BlockRestrictable.h"
#include "INSFVPressureVariable.h"
#include "NonADFunctorInterface.h"

class MooseMesh;
class INSFVPressureVariable;
```

GiudGiud commented on modules/navier_stokes/src/userobjects/NSFVPressurePin.C:
```diff
@@ -0,0 +1,127 @@
+void
+NSFVPressurePin::execute()
+{
+  // For average pressure interpolation, we expect a checkerboard.
+  // We can smooth it here by average each value with the average of several of its neighbors
+  // This will change the value of the pressure pin, but if we are careful it will not change the
+  // average (provided by the postprocessor)
+
```
```suggestion
```

GiudGiud committed:
Address Alex's review

- remove extraneous headers
- remove un-needed interface
- fixup test requirement
- make sure ghosted vector is updated

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>
